integration/data.go

```diff
@@ -97,36 +99,88 @@ func verifySeriesMapForRange(
 	start, end time.Time,
 	namespace string,
 	expected seriesList,
+	expectedDebugFilePath string,
+	actualDebugFilePath string,
 ) {
 	actual := make(seriesList, len(expected))
 	req := rpc.NewFetchRequest()
 	for i := range expected {
 		s := &expected[i]
 		req.NameSpace = namespace
-		req.ID = s.id
+		req.ID = s.ID
 		req.RangeStart = xtime.ToNormalizedTime(start, time.Second)
 		req.RangeEnd = xtime.ToNormalizedTime(end, time.Second)
 		req.ResultTimeType = rpc.TimeType_UNIX_SECONDS
 		fetched, err := ts.fetch(req)
 		require.NoError(t, err)
 		actual[i] = series{
-			id:   s.id,
-			data: fetched,
+			ID:   s.ID,
+			Data: fetched,
 		}
 	}
+
+	if len(expectedDebugFilePath) > 0 {
+		writeVerifyDebugOutput(t, expectedDebugFilePath, start, end, expected)
+	}
+	if len(actualDebugFilePath) > 0 {
+		writeVerifyDebugOutput(t, actualDebugFilePath, start, end, actual)
+	}
+
 	require.Equal(t, expected, actual)
 }
 
+func writeVerifyDebugOutput(t *testing.T, filePath string, start, end time.Time, series seriesList) {
+	w, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY, os.ModeAppend)
+	require.NoError(t, err)
+
+	data, err := json.MarshalIndent(struct {
+		Start  time.Time
+		End    time.Time
+		Series seriesList
+	}{
+		Start:  start,
+		End:    end,
+		Series: series,
+	}, "", "    ")
+	require.NoError(t, err)
+
+	_, err = w.Write(data)
+	require.NoError(t, err)
+	require.NoError(t, w.Close())
+}
+
 func verifySeriesMaps(
 	t *testing.T,
 	ts *testSetup,
 	namespace string,
 	seriesMaps map[time.Time]seriesList,
 ) {
+	var (
+		debugFilePathPrefix   = ts.opts.VerifySeriesDebugFilePathPrefix()
+		expectedDebugFilePath string
+		actualDebugFilePath   string
+	)
+	if debugFilePathPrefix != "" {
+		for _, entry := range []struct {
+			suffix   string
+			filePath *string
+		}{
+			{"expected.log", &expectedDebugFilePath},
+			{"actual.log", &actualDebugFilePath},
+		} {
+			*entry.filePath = debugFilePathPrefix + "_" + entry.suffix
```

Comment:
nit: put the file operations in a function and call it on "expected.log" and "actual.log" separately? not sure if it worth to loop through 2 structs